### PR TITLE
ext: lib: crypto: Extend generic mbedTLS config with HAVE_ASM

### PR DESCRIPTION
**Pull request by aurel32**
Extend generic mbedTLS configuration file with MBEDTLS_HAVE_ASM option,
to allow the use of assembly code. This improves the performances of
asymetric cryptography, however depending on the architecture and the
CPU, this might have an impact on the code size.

Set the default value ot the previous non-configurable value, ie enable
it by default except on ARM.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

### CODE REVIEW

**Comment by aurel32** [comment]
### Benchmark on Cortex-M7 (SAM E70 Xplained)

#### MBEDTLS_HAVE_ASM enabled

```
Memory region         Used Size  Region Size  %age Used
           FLASH:      113072 B         2 MB      5.39%
            SRAM:       87892 B       384 KB     22.35%
        IDT_LIST:          24 B         2 KB      1.17%

***** Booting Zephyr OS zephyr-v1.13.0-3345-g2c8d892f52 *****
        MBEDTLS Benchmark sample
  MD4                      :      60660 KiB/s,         16 ns/byte
  MD5                      :      39619 KiB/s,         24 ns/byte
  SHA-1                    :      23015 KiB/s,         42 ns/byte
  SHA-256                  :      10361 KiB/s,         95 ns/byte
  SHA-512                  :       3216 KiB/s,        305 ns/byte
  ARC4                     :      16230 KiB/s,         60 ns/byte
  3DES                     :       1295 KiB/s,        755 ns/byte
  DES                      :       3346 KiB/s,        292 ns/byte
  AES-CBC-128              :       5960 KiB/s,        165 ns/byte
  AES-CBC-192              :       5148 KiB/s,        189 ns/byte
  AES-CBC-256              :       4571 KiB/s,        214 ns/byte
  AES-XTS-128              :       5253 KiB/s,        187 ns/byte
  AES-XTS-256              :       4134 KiB/s,        238 ns/byte
  AES-GCM-128              :       2416 KiB/s,        406 ns/byte
  AES-GCM-192              :       2269 KiB/s,        433 ns/byte
  AES-GCM-256              :       2160 KiB/s,        456 ns/byte
  AES-CCM-128              :       2582 KiB/s,        380 ns/byte
  AES-CCM-192              :       2280 KiB/s,        430 ns/byte
  AES-CCM-256              :       2036 KiB/s,        481 ns/byte
  ChaCha20-Poly1305        :       3952 KiB/s,        249 ns/byte
  CAMELLIA-CBC-128         :       3799 KiB/s,        258 ns/byte
  CAMELLIA-CBC-192         :       2950 KiB/s,        331 ns/byte
  CAMELLIA-CBC-256         :       2965 KiB/s,        331 ns/byte
  ChaCha20                 :       5064 KiB/s,        194 ns/byte
  Poly1305                 :      26524 KiB/s,         36 ns/byte
  BLOWFISH-CBC-128         :       5209 KiB/s,        188 ns/byte
  BLOWFISH-CBC-192         :       5213 KiB/s,        188 ns/byte
  BLOWFISH-CBC-256         :       5216 KiB/s,        188 ns/byte
  CTR_DRBG (NOPR)          :       4480 KiB/s,        219 ns/byte
  CTR_DRBG (PR)            :       3131 KiB/s,        313 ns/byte
  HMAC_DRBG SHA-1 (NOPR)   :       1460 KiB/s,        673 ns/byte
  HMAC_DRBG SHA-1 (PR)     :       1340 KiB/s,        734 ns/byte
  HMAC_DRBG SHA-256 (NOPR) :        998 KiB/s,        983 ns/byte
  HMAC_DRBG SHA-256 (PR)   :        995 KiB/s,        985 ns/byte
  RSA-2048                 :     155  public/s
  RSA-2048                 :       4 private/s
  RSA-4096                 :      41  public/s
  RSA-4096                 :       1 private/s
  DHE-2048                 :       1 handshake/s
  DH-2048                  :       1 handshake/s
  DHE-3072                 :       0 handshake/s
  DH-3072                  :       0 handshake/s
  ECDSA-secp521r1          :      15 sign/s
  ECDSA-brainpoolP512r1    :       1 sign/s
  ECDSA-secp384r1          :      22 sign/s
  ECDSA-brainpoolP384r1    :       2 sign/s
  ECDSA-secp256r1          :      34 sign/s
  ECDSA-secp256k1          :      36 sign/s
  ECDSA-brainpoolP256r1    :       5 sign/s
  ECDSA-secp224r1          :      52 sign/s
  ECDSA-secp224k1          :      43 sign/s
  ECDSA-secp192r1          :      70 sign/s
  ECDSA-secp192k1          :      53 sign/s
  ECDSA-secp521r1          :       4 verify/s
  ECDSA-brainpoolP512r1    :       0 verify/s
  ECDSA-secp384r1          :       6 verify/s
  ECDSA-brainpoolP384r1    :       1 verify/s
  ECDSA-secp256r1          :       9 verify/s
  ECDSA-secp256k1          :      11 verify/s
  ECDSA-brainpoolP256r1    :       1 verify/s
  ECDSA-secp224r1          :      15 verify/s
  ECDSA-secp224k1          :      12 verify/s
  ECDSA-secp192r1          :      21 verify/s
  ECDSA-secp192k1          :      15 verify/s
  ECDHE-secp521r1          :       4 handshake/s
  ECDHE-brainpoolP512r1    :       0 handshake/s
  ECDHE-secp384r1          :       6 handshake/s
  ECDHE-brainpoolP384r1    :       1 handshake/s
  ECDHE-secp256r1          :      10 handshake/s
  ECDHE-secp256k1          :      11 handshake/s
  ECDHE-brainpoolP256r1    :       1 handshake/s
  ECDHE-secp224r1          :      16 handshake/s
  ECDHE-secp224k1          :      13 handshake/s
  ECDHE-secp192r1          :      22 handshake/s
  ECDHE-secp192k1          :      16 handshake/s
  ECDHE-Curve25519         :       9 handshake/s
  ECDHE-Curve448           :       4 handshake/s
  ECDH-secp521r1           :       6 handshake/s
  ECDH-brainpoolP512r1     :       1 handshake/s
  ECDH-secp384r1           :       8 handshake/s
  ECDH-brainpoolP384r1     :       1 handshake/s
  ECDH-secp256r1           :      14 handshake/s
  ECDH-secp256k1           :      16 handshake/s
  ECDH-brainpoolP256r1     :       2 handshake/s
  ECDH-secp224r1           :      22 handshake/s
  ECDH-secp224k1           :      18 handshake/s
  ECDH-secp192r1           :      30 handshake/s
  ECDH-secp192k1           :      22 handshake/s
  ECDH-Curve25519          :      17 handshake/s
  ECDH-Curve448            :       8 handshake/s

       Done
```

#### MBEDTLS_HAVE_ASM disabled
```

Memory region         Used Size  Region Size  %age Used
           FLASH:      113936 B         2 MB      5.43%
            SRAM:       87892 B       384 KB     22.35%
        IDT_LIST:          24 B         2 KB      1.17%

***** Booting Zephyr OS zephyr-v1.13.0-3345-g2c8d892f52 *****
        MBEDTLS Benchmark sample
  MD4                      :      60703 KiB/s,         16 ns/byte
  MD5                      :      39591 KiB/s,         24 ns/byte
  SHA-1                    :      23059 KiB/s,         42 ns/byte
  SHA-256                  :      10365 KiB/s,         95 ns/byte
  SHA-512                  :       3216 KiB/s,        305 ns/byte
  ARC4                     :      16230 KiB/s,         60 ns/byte
  3DES                     :       1294 KiB/s,        756 ns/byte
  DES                      :       3327 KiB/s,        293 ns/byte
  AES-CBC-128              :       5954 KiB/s,        165 ns/byte
  AES-CBC-192              :       5152 KiB/s,        189 ns/byte
  AES-CBC-256              :       4575 KiB/s,        214 ns/byte
  AES-XTS-128              :       5254 KiB/s,        187 ns/byte
  AES-XTS-256              :       4135 KiB/s,        238 ns/byte
  AES-GCM-128              :       2438 KiB/s,        402 ns/byte
  AES-GCM-192              :       2288 KiB/s,        427 ns/byte
  AES-GCM-256              :       2169 KiB/s,        453 ns/byte
  AES-CCM-128              :       2590 KiB/s,        380 ns/byte
  AES-CCM-192              :       2279 KiB/s,        431 ns/byte
  AES-CCM-256              :       2033 KiB/s,        481 ns/byte
  ChaCha20-Poly1305        :       3929 KiB/s,        251 ns/byte
  CAMELLIA-CBC-128         :       3776 KiB/s,        259 ns/byte
  CAMELLIA-CBC-192         :       2942 KiB/s,        333 ns/byte
  CAMELLIA-CBC-256         :       2948 KiB/s,        333 ns/byte
  ChaCha20                 :       5068 KiB/s,        193 ns/byte
  Poly1305                 :      26546 KiB/s,         36 ns/byte
  BLOWFISH-CBC-128         :       5212 KiB/s,        188 ns/byte
  BLOWFISH-CBC-192         :       5220 KiB/s,        188 ns/byte
  BLOWFISH-CBC-256         :       5217 KiB/s,        188 ns/byte
  CTR_DRBG (NOPR)          :       4528 KiB/s,        217 ns/byte
  CTR_DRBG (PR)            :       3170 KiB/s,        310 ns/byte
  HMAC_DRBG SHA-1 (NOPR)   :       1463 KiB/s,        675 ns/byte
  HMAC_DRBG SHA-1 (PR)     :       1333 KiB/s,        737 ns/byte
  HMAC_DRBG SHA-256 (NOPR) :        993 KiB/s,        984 ns/byte
  HMAC_DRBG SHA-256 (PR)   :        994 KiB/s,        985 ns/byte
  RSA-2048                 :     385  public/s
  RSA-2048                 :       8 private/s
  RSA-4096                 :     108  public/s
  RSA-4096                 :       1 private/s
  DHE-2048                 :       2 handshake/s
  DH-2048                  :       3 handshake/s
  DHE-3072                 :       1 handshake/s
  DH-3072                  :       1 handshake/s
  ECDSA-secp521r1          :      18 sign/s
  ECDSA-brainpoolP512r1    :       1 sign/s
  ECDSA-secp384r1          :      24 sign/s
  ECDSA-brainpoolP384r1    :       3 sign/s
  ECDSA-secp256r1          :      36 sign/s
  ECDSA-secp256k1          :      39 sign/s
  ECDSA-brainpoolP256r1    :       5 sign/s
  ECDSA-secp224r1          :      56 sign/s
  ECDSA-secp224k1          :      47 sign/s
  ECDSA-secp192r1          :      72 sign/s
  ECDSA-secp192k1          :      56 sign/s
  ECDSA-secp521r1          :       5 verify/s
  ECDSA-brainpoolP512r1    :       0 verify/s
  ECDSA-secp384r1          :       7 verify/s
  ECDSA-brainpoolP384r1    :       1 verify/s
  ECDSA-secp256r1          :      10 verify/s
  ECDSA-secp256k1          :      12 verify/s
  ECDSA-brainpoolP256r1    :       1 verify/s
  ECDSA-secp224r1          :      16 verify/s
  ECDSA-secp224k1          :      13 verify/s
  ECDSA-secp192r1          :      22 verify/s
  ECDSA-secp192k1          :      16 verify/s
  ECDHE-secp521r1          :       6 handshake/s
  ECDHE-brainpoolP512r1    :       0 handshake/s
  ECDHE-secp384r1          :       7 handshake/s
  ECDHE-brainpoolP384r1    :       1 handshake/s
  ECDHE-secp256r1          :      11 handshake/s
  ECDHE-secp256k1          :      13 handshake/s
  ECDHE-brainpoolP256r1    :       1 handshake/s
  ECDHE-secp224r1          :      18 handshake/s
  ECDHE-secp224k1          :      14 handshake/s
  ECDHE-secp192r1          :      23 handshake/s
  ECDHE-secp192k1          :      17 handshake/s
  ECDHE-Curve25519         :       9 handshake/s
  ECDHE-Curve448           :       5 handshake/s
  ECDH-secp521r1           :       7 handshake/s
  ECDH-brainpoolP512r1     :       1 handshake/s
  ECDH-secp384r1           :       9 handshake/s
  ECDH-brainpoolP384r1     :       1 handshake/s
  ECDH-secp256r1           :      15 handshake/s
  ECDH-secp256k1           :      17 handshake/s
  ECDH-brainpoolP256r1     :       2 handshake/s
  ECDH-secp224r1           :      24 handshake/s
  ECDH-secp224k1           :      19 handshake/s
  ECDH-secp192r1           :      31 handshake/s
  ECDH-secp192k1           :      24 handshake/s
  ECDH-Curve25519          :      18 handshake/s
  ECDH-Curve448            :      10 handshake/s

       Done
```

**Comment by nashif** [comment]
which one is "MBEDTLS_HAVE_ASM disabled" ? looks like you have both as disabled.

**Comment by aurel32** [comment]
> which one is "MBEDTLS_HAVE_ASM disabled" ? looks like you have both as disabled.

Indeed, I mixed-up things during copy and paste, I have just fixed that. The first one is disabled, and the second enabled. When enabled, it's smaller and faster for asymmetric cryptography (mostly RSA, a bit ECC), in the noise for hash and encryption.

I have benchmark running on Cortex-M4 (nRF52840 PCA10056), I'll try to be more careful with copy and paste.

**Comment by aurel32** [comment]
### Benchmark on Cortex-M4 (nRF52840 PCA10056)

Note that this is with PR#12619 applied.

#### MBEDTLS_HAVE_ASM disabled

```
Memory region         Used Size  Region Size  %age Used
           FLASH:      145252 B         1 MB     13.85%
            SRAM:       85052 B       256 KB     32.44%
        IDT_LIST:         104 B         2 KB      5.08%

***** Booting Zephyr OS zephyr-v1.13.0-3345-g2c8d892f52 *****
        MBEDTLS Benchmark sample
  MD4                      :       7033 KiB/s,        140 ns/byte
  MD5                      :       4568 KiB/s,        214 ns/byte
  SHA-1                    :       1992 KiB/s,        493 ns/byte
  SHA-256                  :       1184 KiB/s,        827 ns/byte
  SHA-512                  :        329 KiB/s,       2990 ns/byte
  ARC4                     :       2239 KiB/s,        424 ns/byte
  3DES                     :        117 KiB/s,        304 ns/byte
  DES                      :        312 KiB/s,       3137 ns/byte
  AES-CBC-128              :        537 KiB/s,       1827 ns/byte
  AES-CBC-192              :        461 KiB/s,       2126 ns/byte
  AES-CBC-256              :        405 KiB/s,       2425 ns/byte
  AES-XTS-128              :        481 KiB/s,       2037 ns/byte
  AES-XTS-256              :        370 KiB/s,       2646 ns/byte
  AES-GCM-128              :        224 KiB/s,        293 ns/byte
  AES-GCM-192              :        211 KiB/s,        601 ns/byte
  AES-GCM-256              :        197 KiB/s,        909 ns/byte
  AES-CCM-128              :        241 KiB/s,       4072 ns/byte
  AES-CCM-192              :        210 KiB/s,        593 ns/byte
  AES-CCM-256              :        186 KiB/s,       1203 ns/byte
  ChaCha20-Poly1305        :        491 KiB/s,       2000 ns/byte
  CAMELLIA-CBC-128         :        428 KiB/s,       2287 ns/byte
  CAMELLIA-CBC-192         :        334 KiB/s,       2953 ns/byte
  CAMELLIA-CBC-256         :        335 KiB/s,       2952 ns/byte
  ChaCha20                 :        616 KiB/s,       1600 ns/byte
  Poly1305                 :       3259 KiB/s,        299 ns/byte
  BLOWFISH-CBC-128         :        753 KiB/s,       1302 ns/byte
  BLOWFISH-CBC-192         :        750 KiB/s,       1303 ns/byte
  BLOWFISH-CBC-256         :        756 KiB/s,       1303 ns/byte
  CTR_DRBG (NOPR)          :        401 KiB/s,       2459 ns/byte
  CTR_DRBG (PR)            :        125 KiB/s,       3813 ns/byte
  HMAC_DRBG SHA-1 (NOPR)   :        135 KiB/s,       3267 ns/byte
  HMAC_DRBG SHA-1 (PR)     :        109 KiB/s,        917 ns/byte
  HMAC_DRBG SHA-256 (NOPR) :         88 KiB/s,       3112 ns/byte
  HMAC_DRBG SHA-256 (PR)   :         88 KiB/s,       3122 ns/byte
  RSA-2048                 :      19  public/s
  RSA-2048                 :       1 private/s
  RSA-4096                 :       5  public/s
  RSA-4096                 :       0 private/s
  DHE-2048                 :       0 handshake/s
  DH-2048                  :       0 handshake/s
  DHE-3072                 :       0 handshake/s
  DH-3072                  :       0 handshake/s
  ECDSA-secp521r1          :       1 sign/s
  ECDSA-brainpoolP512r1    :       0 sign/s
  ECDSA-secp384r1          :       2 sign/s
  ECDSA-brainpoolP384r1    :       0 sign/s
  ECDSA-secp256r1          :       4 sign/s
  ECDSA-secp256k1          :       4 sign/s
  ECDSA-brainpoolP256r1    :       0 sign/s
  ECDSA-secp224r1          :       7 sign/s
  ECDSA-secp224k1          :       5 sign/s
  ECDSA-secp192r1          :       9 sign/s
  ECDSA-secp192k1          :       7 sign/s
  ECDSA-secp521r1          :       1 verify/s
  ECDSA-brainpoolP512r1    :       0 verify/s
  ECDSA-secp384r1          :       1 verify/s
  ECDSA-brainpoolP384r1    :       0 verify/s
  ECDSA-secp256r1          :       1 verify/s
  ECDSA-secp256k1          :       2 verify/s
  ECDSA-brainpoolP256r1    :       0 verify/s
  ECDSA-secp224r1          :       2 verify/s
  ECDSA-secp224k1          :       2 verify/s
  ECDSA-secp192r1          :       3 verify/s
  ECDSA-secp192k1          :       2 verify/s
  ECDHE-secp521r1          :       1 handshake/s
  ECDHE-brainpoolP512r1    :       0 handshake/s
  ECDHE-secp384r1          :       1 handshake/s
  ECDHE-brainpoolP384r1    :       0 handshake/s
  ECDHE-secp256r1          :       1 handshake/s
  ECDHE-secp256k1          :       2 handshake/s
  ECDHE-brainpoolP256r1    :       0 handshake/s
  ECDHE-secp224r1          :       2 handshake/s
  ECDHE-secp224k1          :       2 handshake/s
  ECDHE-secp192r1          :       3 handshake/s
  ECDHE-secp192k1          :       2 handshake/s
  ECDHE-Curve25519         :       1 handshake/s
  ECDHE-Curve448           :       1 handshake/s
  ECDH-secp521r1           :       1 handshake/s
  ECDH-brainpoolP512r1     :       0 handshake/s
  ECDH-secp384r1           :       1 handshake/s
  ECDH-brainpoolP384r1     :       0 handshake/s
  ECDH-secp256r1           :       2 handshake/s
  ECDH-secp256k1           :       2 handshake/s
  ECDH-brainpoolP256r1     :       0 handshake/s
  ECDH-secp224r1           :       3 handshake/s
  ECDH-secp224k1           :       3 handshake/s
  ECDH-secp192r1           :       4 handshake/s
  ECDH-secp192k1           :       3 handshake/s
  ECDH-Curve25519          :       2 handshake/s
  ECDH-Curve448            :       1 handshake/s

       Done

```
#### MBEDTLS_HAVE_ASM enabled

```
Memory region         Used Size  Region Size  %age Used
           FLASH:      144420 B         1 MB     13.77%
            SRAM:       85052 B       256 KB     32.44%
        IDT_LIST:         104 B         2 KB      5.08%

***** Booting Zephyr OS zephyr-v1.13.0-3345-g2c8d892f52 *****
        MBEDTLS Benchmark sample
  MD4                      :       7042 KiB/s,        139 ns/byte
  MD5                      :       4577 KiB/s,        214 ns/byte
  SHA-1                    :       1994 KiB/s,        493 ns/byte
  SHA-256                  :       1184 KiB/s,        828 ns/byte
  SHA-512                  :        329 KiB/s,       2991 ns/byte
  ARC4                     :       2233 KiB/s,        424 ns/byte
  3DES                     :        118 KiB/s,        255 ns/byte
  DES                      :        307 KiB/s,       3184 ns/byte
  AES-CBC-128              :        538 KiB/s,       1827 ns/byte
  AES-CBC-192              :        460 KiB/s,       2128 ns/byte
  AES-CBC-256              :        405 KiB/s,       2428 ns/byte
  AES-XTS-128              :        478 KiB/s,       2041 ns/byte
  AES-XTS-256              :        372 KiB/s,       2650 ns/byte
  AES-GCM-128              :        229 KiB/s,        219 ns/byte
  AES-GCM-192              :        214 KiB/s,        525 ns/byte
  AES-GCM-256              :        200 KiB/s,        830 ns/byte
  AES-CCM-128              :        243 KiB/s,       4059 ns/byte
  AES-CCM-192              :        212 KiB/s,        576 ns/byte
  AES-CCM-256              :        186 KiB/s,       1186 ns/byte
  ChaCha20-Poly1305        :        490 KiB/s,       1997 ns/byte
  CAMELLIA-CBC-128         :        426 KiB/s,       2304 ns/byte
  CAMELLIA-CBC-192         :        330 KiB/s,       2969 ns/byte
  CAMELLIA-CBC-256         :        331 KiB/s,       2969 ns/byte
  ChaCha20                 :        613 KiB/s,       1596 ns/byte
  Poly1305                 :       3269 KiB/s,        300 ns/byte
  BLOWFISH-CBC-128         :        749 KiB/s,       1309 ns/byte
  BLOWFISH-CBC-192         :        749 KiB/s,       1309 ns/byte
  BLOWFISH-CBC-256         :        749 KiB/s,       1309 ns/byte
  CTR_DRBG (NOPR)          :        403 KiB/s,       2431 ns/byte
  CTR_DRBG (PR)            :        126 KiB/s,       3780 ns/byte
  HMAC_DRBG SHA-1 (NOPR)   :        135 KiB/s,       3240 ns/byte
  HMAC_DRBG SHA-1 (PR)     :        110 KiB/s,        885 ns/byte
  HMAC_DRBG SHA-256 (NOPR) :         88 KiB/s,       3128 ns/byte
  HMAC_DRBG SHA-256 (PR)   :         88 KiB/s,       3128 ns/byte
  RSA-2048                 :      50  public/s
  RSA-2048                 :       1 private/s
  RSA-4096                 :      14  public/s
  RSA-4096                 :       0 private/s
  DHE-2048                 :       0 handshake/s
  DH-2048                  :       0 handshake/s
  DHE-3072                 :       0 handshake/s
  DH-3072                  :       0 handshake/s
  ECDSA-secp521r1          :       2 sign/s
  ECDSA-brainpoolP512r1    :       0 sign/s
  ECDSA-secp384r1          :       3 sign/s
  ECDSA-brainpoolP384r1    :       0 sign/s
  ECDSA-secp256r1          :       4 sign/s
  ECDSA-secp256k1          :       5 sign/s
  ECDSA-brainpoolP256r1    :       0 sign/s
  ECDSA-secp224r1          :       7 sign/s
  ECDSA-secp224k1          :       6 sign/s
  ECDSA-secp192r1          :      10 sign/s
  ECDSA-secp192k1          :       7 sign/s
  ECDSA-secp521r1          :       1 verify/s
  ECDSA-brainpoolP512r1    :       0 verify/s
  ECDSA-secp384r1          :       1 verify/s
  ECDSA-brainpoolP384r1    :       0 verify/s
  ECDSA-secp256r1          :       1 verify/s
  ECDSA-secp256k1          :       2 verify/s
  ECDSA-brainpoolP256r1    :       0 verify/s
  ECDSA-secp224r1          :       2 verify/s
  ECDSA-secp224k1          :       2 verify/s
  ECDSA-secp192r1          :       3 verify/s
  ECDSA-secp192k1          :       2 verify/s
  ECDHE-secp521r1          :       1 handshake/s
  ECDHE-brainpoolP512r1    :       0 handshake/s
  ECDHE-secp384r1          :       1 handshake/s
  ECDHE-brainpoolP384r1    :       0 handshake/s
  ECDHE-secp256r1          :       1 handshake/s
  ECDHE-secp256k1          :       2 handshake/s
  ECDHE-brainpoolP256r1    :       0 handshake/s
  ECDHE-secp224r1          :       2 handshake/s
  ECDHE-secp224k1          :       2 handshake/s
  ECDHE-secp192r1          :       3 handshake/s
  ECDHE-secp192k1          :       2 handshake/s
  ECDHE-Curve25519         :       1 handshake/s
  ECDHE-Curve448           :       1 handshake/s
  ECDH-secp521r1           :       1 handshake/s
  ECDH-brainpoolP512r1     :       0 handshake/s
  ECDH-secp384r1           :       1 handshake/s
  ECDH-brainpoolP384r1     :       0 handshake/s
  ECDH-secp256r1           :       2 handshake/s
  ECDH-secp256k1           :       3 handshake/s
  ECDH-brainpoolP256r1     :       0 handshake/s
  ECDH-secp224r1           :       3 handshake/s
  ECDH-secp224k1           :       3 handshake/s
  ECDH-secp192r1           :       5 handshake/s
  ECDH-secp192k1           :       4 handshake/s
  ECDH-Curve25519          :       3 handshake/s
  ECDH-Curve448            :       1 handshake/s

       Done
```

**Comment by rlubos** [comment]
Why is it disabled by default on ARM? I can see that in both cases you present, with `MBEDTLS_HAVE_ASM ` enabled, code size is smaller and computations are faster (looking at RSA).

**Comment by aurel32** [comment]
> Why is it disabled by default on ARM? I can see that in both cases you present, with `MBEDTLS_HAVE_ASM ` enabled, code size is smaller and computations are faster (looking at RSA).

I don't really know, the initial version of the config files already have this option disabled on ARM, that is why I preferred not to change the default. One hypothesis is that the impact on the code size is different depending on the ARM core:
- Cortex M0 +616 bytes
- Cortex M3 -184 bytes
- Cortex M4 -832 bytes
- Cortex M7 -864 bytes

The gain in size and performance on Cortex M4/M7 is relatively recent and due to this commit:
https://github.com/ARMmbed/mbedtls/commit/16b1bd89326ece62712c8ecc1142a41bd257d443

**Review by rlubos — APPROVED** [review]
LGTM 👍

File: ext/lib/crypto/mbedtls/Kconfig.tls-generic [hunk 1]
```diff
@@ -266,6 +266,14 @@ config MBEDTLS_PEM_CERTIFICATE_FORMAT
 	  By default only DER (binary) format of certificates is supported. Enable
 	  this option to enable support for PEM format.
 
+config MBEDTLS_HAVE_ASM
+	bool "Enable use of assembly code"
+	default y if !ARM
```
[the remark this answers is not in the document]
Comment (aurel32) [reply ext/lib/crypto/mbedtls/Kconfig.tls-generic]: Yes this PR doesn't change the "default" value. It was disabled for ARM, now it is configurable through Kconfig, but still disabled for ARM. The benchmarks are there to show this option can bring a significant performance improvement, and thus that's a good idea to make it configurable.

**Review by nashif — APPROVED** [review]

**Review by mike-scott — APPROVED** [review]

**Pull request merged**